README.rst
Chandler
========

A simple CLI utility to display informations about OAR cluster from the API.


Installation
------------

Requirements:
  - python >= 2.7

You can install, upgrade, uninstall chandler with these commands::

  $ pip install [--user] chandler
  $ pip install [--user] --upgrade chandler
  $ pip uninstall chandler

Or from git (last development version)::

  $ pip install git+https://github.com/oar-team/chandler.git

Or if you already pulled the sources::

  $ pip install path/to/sources

Or if you don't have pip::

  $ easy_install chandler


CHANGES
Chandler
========

Version 0.1.1.dev0
------------------

**unreleased**


Version 0.1.0
-------------

Released on April 29th 2015

Initial public release


chandler/main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import absolute_import, unicode_literals

from natsort import natsorted
import sys
import os
import time
import re
from optparse import OptionParser
from collections import defaultdict
from colorama import Fore, Back, Style

from .configuration import Configuration
from .utils import get_resources, get_jobs, cprint, init_colorama, reraise


def run(options, args):
    config = Configuration()

    # Get some variables from the configuration file
    COLS = config.getint('output', 'columns')
    NODENAME_REGEX = config.get('output', 'nodename_regex')
    COL_SIZE = config.getint('output', 'col_size')
    COL_SPAN = config.getint('output', 'col_span')
    MAX_COLS = config.getint('output', 'max_cols')
    USERS_STATS_BY_DEFAULT = config.getboolean('output', 'users_stats_by_default')
    NODES_USAGE_BY_DEFAULT = config.getboolean('output', 'nodes_usage_by_default')
    NODES_HEADER = config.get('output', 'nodes_header')
    NODES_FORMAT = config.get('output', 'nodes_format')
    try:
        COMMENT_PROPERTY = config.get('output', 'comment_property')
    except:
        COMMENT_PROPERTY = ""
    try:
        SEPARATIONS = config.get('output', 'separations')
    except:
        SEPARATIONS = ""
    SEPARATIONS = SEPARATIONS.split(',')

    # Compute the number of columns depending on the COLUMNS environment variable
    try:
        rows, columns = os.popen('stty size', 'r').read().split()
        COLS = int(int(columns) / COL_SIZE)
    except:
        pass
    if COLS == 0:
        COLS = 1
    if COLS > MAX_COLS:
        COLS = MAX_COLS

    init_colorama()
    # Get rid of http_proxy if necessary
    if config.getboolean('misc', 'ignore_proxy'):
        try:
            del os.environ['http_proxy']
        except:
            pass
        try:
            del os.environ['https_proxy']
        except:
            pass

    # Print a waiting message
    print('Querying OAR API...\n\033[1A'),

    # Get the data from the API
    # TODO: paginated results management
    resources = get_resources()
    jobs = get_jobs()

    # Erase the waiting message
    print("\033[2K")

    # Compute assigned resources dictionnary
    assigned_resources = {r["id"]: j["types"] for j in jobs
                          for r in j["resources"]
                          if r["status"] == "assigned"}

    # Compute sorted node list
    nodes = natsorted(set([r["network_address"] for r in resources]))

    # Get the comment property if necessary
    if COMMENT_PROPERTY != '':
        comment = {r["network_address"]: r[COMMENT_PROPERTY] for r in resources}

    # Loop on nodes and resources
    col = 0
    down = 0
    for node in nodes:
        c = 0
        node_resources = [r for r in resources if r["network_address"] == node]
        p = re.match(NODENAME_REGEX, node)
        node_str = p.group(1)
        if COMMENT_PROPERTY != '':
            node_str += " (" + comment[node] + ")"
        else:
            node_str += ": "
        string = node_str + " " * (COL_SPAN - len(node_str))
        cprint(Fore.RESET + Back.RESET + string)
        for r in node_resources:
            c += 1
            if r["state"] == "Dead":
                down += 1
                cprint(Back.RED + Fore.WHITE + "D")
            elif r["state"] == "Absent":
                if int(r["available_upto"]) > time.time():
                    cprint(Back.CYAN + Fore.WHITE + " ")
                else:
                    down += 1
                    cprint(Back.RED + Fore.WHITE + "A")
            elif r["state"] == "Suspected":
                down += 1
                cprint(Back.RED + Fore.WHITE + "S")
            elif r["state"] == "Alive":
                try:
                    types = assigned_resources[r["id"]]
                except:
                    cprint(Back.GREEN + Fore.WHITE + " ")
                else:
                    if "besteffort" in types:
                        cprint(Back.GREEN + Fore.BLACK + "B")
                    elif "timesharing" in types:
                        cprint(Back.YELLOW + Fore.BLACK + "T")
                    else:
                        cprint(Back.WHITE + Fore.BLACK + "J")
        cprint(Fore.RESET + Back.RESET)
        col += 1
        if col < COLS and node not in SEPARATIONS:
            cprint(" " * (COL_SIZE - COL_SPAN - c))
        else:
            col = 0
            print

    # Legend
    print(Fore.RESET + Back.RESET)
    print
    cprint(Back.GREEN + " " + Back.RESET + "=Free ")
    cprint(Back.GREEN + Fore.BLACK + "B" +
           Back.RESET + Fore.RESET + "=Besteffort ")
    cprint(Back.CYAN + " " + Back.RESET + "=Standby ")
    cprint(Back.WHITE + Fore.BLACK + "J" + Back.RESET + Fore.RESET + "=Job ")
    cprint(Back.RED + Fore.BLACK + "S" + Back.RESET + Fore.RESET + "=Suspected ")
    cprint(Back.RED + Fore.BLACK + "A" + Back.RESET + Fore.RESET + "=Absent ")
    cprint(Back.RED + Fore.BLACK + "D" + Back.RESET + Fore.RESET + "=Dead ")
    print

    # Reset terminal styles
    print(Fore.RESET + Back.RESET + Style.RESET_ALL)

    # Print summary
    print "{} jobs, {} resources, {} down, {} used".format(len(jobs),
                                                           len(resources),
                                                           down,
                                                           len(assigned_resources))

    # Print users stats if necessary
    if USERS_STATS_BY_DEFAULT ^ options.toggle_users and len(jobs) > 0:
        print
        user_resources = defaultdict(int)
        user_running = defaultdict(int)
        user_waiting = defaultdict(int)
        user_nodes = defaultdict(list)
        for j in jobs:
            if (j["state"] == "Running" or j["state"] == "Finishing"
                    or j["state"] == "Launching"):
                user_running[j["owner"]] += 1
                user_resources[j["owner"]] += len(j["resources"])
                user_nodes[j["owner"]] += [n["network_address"]
                                           for n in j["nodes"]]
            elif j["state"] == "Waiting":
                user_waiting[j["owner"]] += 1
                user_resources[j["owner"]] += 0
                user_nodes[j["owner"]] += []

        print "               Jobs       Jobs"
        print "User          running    waiting   Resources    Nodes"
        print "====================================================="
        for u, r in user_resources.iteritems():
            nodes = set(user_nodes[u])
            print "{:<16} {:<10} {:<10} {:<10} {:<10}".format(u, user_running[u],
                                                              user_waiting[u],
                                                              r, len(nodes))

    # Print nodes usage if necessary
    if NODES_USAGE_BY_DEFAULT ^ options.toggle_nodes and len(jobs) > 0:
        print
        assigned_nodes = [[r["network_address"], j]
                          for j in jobs
                          for r in j["nodes"] if r["status"] == "assigned"]
        nodes_usage = defaultdict(list)
        for c in assigned_nodes:
            nodes_usage[c[0]].append(c[1])
        print NODES_HEADER
        for node in natsorted(nodes_usage.keys()):
            print node + ":"
            for job in nodes_usage[node]:
                remain_time = job["start_time"] + job["walltime"] - time.time()
                d = time.strftime("%H:%M:%S", time.gmtime(remain_time))
                r = [r for r in resources if r["network_address"] ==
                     node and r["id"] in [rj["id"] for rj in job["resources"]]]
                print "    " + NODES_FORMAT.format(job["id"], job["owner"],
                                                   str(job["name"]), len(r),
                                                   ",".join(job["types"]),
                                                   d, job["project"])


def main():
    # Options parsing
    parser = OptionParser()
    parser.add_option("-u", "--users",
                      action="store_true", dest="toggle_users", default=False,
                      help="Toggle printing users stats")
    parser.add_option("-n", "--nodes",
                      action="store_true", dest="toggle_nodes", default=False,
                      help="Toggle printing nodes usage")
    parser.add_option("-r", "--reload-cache",
                      action="store_true", dest="reload_cache", default=False,
                      help="Reload the cache")
    parser.add_option("--debug",
                      action="store_true", dest="debug", default=False,
                      help="Enable debug mode")
    (options, args) = parser.parse_args()

    try:
        run(options, args)
    except:
        exc_type, exc_value, tb = sys.exc_info()
        if not options.debug:
            sys.stderr.write("%s\n" % exc_value)
            sys.exit(1)
        else:
            reraise(exc_type, exc_value, tb.tb_next)


chandler/utils.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

import os
import time
import sys
from io import open
try:
    import simplejson as json
except ImportError:
    import json
import requests
from distutils.version import StrictVersion

PY3 = sys.version_info[0] == 3


if PY3:
    def reraise(tp, value, tb=None):
        if value.__traceback__ is not tb:
            raise value.with_traceback(tb)
        raise value

else:
    exec('def reraise(tp, value, tb=None):\n raise tp, value, tb')


def touch(fname, times=None):
    dirname = '/'.join(fname.split('/')[:-1])
    if not os.path.exists(dirname):
        os.makedirs(dirname)
    with file(fname, 'a'):
        os.utime(fname, times)


def get_from_api(uri, config):
    """
        Get an object from the api
    """
    api_uri = config.get('oarapi', 'uri')
    api_limit = config.get('oarapi', 'limit')
    headers = {'Accept': 'application/json'}
    r = requests.get(api_uri + uri + "?limit=" + api_limit, headers=headers)
    if r.status_code != 200:
        print("Could not get " + api_uri + uri)
        r.raise_for_status()
    if StrictVersion(requests.__version__) >= StrictVersion("2.0.0"):
        return r.json()
    else:
        return r.json


def get(uri, config, reload_cache=False):
    """
        Get from the cache or from the api
    """
    view = uri.split("/")[1]
    if (config.cache['enabled'] and not reload_cache):
        cache_time = time.time() - os.path.getmtime(config.cache[view]['file'])
        if (os.path.isfile(config.cache[view]['file'])
                and cache_time < config.cache[view]['delay']):
            with open(config.cache[view]['file']) as fd:
                return json.load(fd)
    else:
        data = get_from_api(uri, config)
    if config.cache['enabled']:
        touch(config.cache[view]['file'])
        with open(config.cache[view]['file'], 'w') as fd:
            json.dump(data, fd)
    return data


def cprint(str, *args):
    """
        Custom print function to get rid of trailing newline and space
    """
    sys.stdout.write(str % args)


def init_colorama():
    from colorama import init
    init()


chandler/configuration.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import, unicode_literals

import os
import pprint
import ConfigParser

from .utils import cached_property


class Configuration(ConfigParser.ConfigParser):
    DEFAULT_CONFIG_FILES = [
        os.path.join(os.environ['HOME'], '.config', 'chandler.conf'),
        os.path.join('/etc', 'oar', 'chandler.conf'),
    ]

    def __init__(self):
        if os.environ.get('CHANDLER_CONF_FILE', None):
            env_file = os.environ['CHANDLER_CONF_FILE']
            self.DEFAULT_CONFIG_FILES.insert(0, env_file)
        ConfigParser.ConfigParser.__init__(self, allow_no_value=False)
        for config_file in self.DEFAULT_CONFIG_FILES[:-1]:
            if self.load_file(config_file, silent=True):
                break
        else:
            self.load_file(self.DEFAULT_CONFIG_FILES[-1], silent=False)

    def load_file(self, filename, silent=False):
        """Updates the values in the config from a config file.
        :param filename: the filename of the config.  This can either be an
                         absolute filename or a filename relative to the
                         root path.
        """
        try:
            with open(filename) as f:
                self.readfp(f)
        except IOError as e:
            if silent:
                return False
            e.strerror = ('Unable to load configuration file. \n\n'
                          'The configuration file is searched into %s or in '
                          'the location given by the $CHANDLER_CONF_FILE '
                          'environement variable' % self.DEFAULT_CONFIG_FILES)
            raise
        return True

    @cached_property
    def nodename_regex(self):
        return self.get('output', 'nodename_regex')

    @cached_property
    def col_size(self):
        return self.getint('output', 'col_size')

    @cached_property
    def col_span(self):
        return self.getint('output', 'col_span')

    @cached_property
    def max_cols(self):
        return self.getint('output', 'max_cols')

    @cached_property
    def users_stats_by_default(self):
        return self.getboolean('output', 'users_stats_by_default')

    @cached_property
    def nodes_usage_by_default(self):
        return self.getboolean('output', 'nodes_usage_by_default')

    @cached_property
    def nodes_header(self):
        return self.get('output', 'nodes_header')

    @cached_property
    def nodes_format(self):
        return self.get('output', 'nodes_format')

    @cached_property
    def comment_property(self):
        try:
            return self.get('output', 'comment_property')
        except:
            pass

    @cached_property
    def separations(self):
        try:
            return self.get('output', 'separations').split(',')
        except:
            return []

    @cached_property
    def cols(self):
        cols = self.getint('output', 'columns')
        # Compute the number of columns depending on the COLUMNS environment
        # variable
        try:
            rows, columns = os.popen('stty size', 'r').read().split()
            cols = int(int(columns) / self.col_size)
        except:
            pass
        if cols == 0:
            cols = 1
        if cols > self.max_cols:
            cols = self.max_cols
        return cols

    @cached_property
    def ignore_proxy(self):
        return self.getboolean('misc', 'ignore_proxy')

    @cached_property
    def cache(self):
        info = {
            'resources': {},
            'jobs': {},
        }
        info['enabled'] = self.getboolean('oarapi', 'caching')
        if info['enabled']:
            for view in ('resources', 'jobs'):
                cache_file = self.get('oarapi', 'caching_%s_file' % view)
                info[view] = {
                    'file': os.path.expanduser(cache_file),
                    'delay': self.getint('oarapi', 'caching_%s_delay' % view)
                }
        return info

    def __str__(self):
        return pprint.pprint(self._sections)


chandler.py
#!/usr/bin/env python
import requests
import json
from natsort import natsorted
import sys
import os
import time
import re
from colorama import init, Fore, Back, Style
init()
import ConfigParser
from optparse import OptionParser
from collections import defaultdict
from distutils.version import StrictVersion

# Configuration file opening
config = ConfigParser.ConfigParser()
DEFAULT_CONFIG_FILE = "/etc/oar/chandler.conf"
try:
    if not os.path.isfile(os.environ['CHANDLER_CONF_FILE']):
        raise
except:
    if os.path.isfile(DEFAULT_CONFIG_FILE):
        config.read(DEFAULT_CONFIG_FILE)
    else:
        sys.stderr.write("Error: could not load configuration file!\n")
        sys.stderr.write("The configuration file is searched into " + DEFAULT_CONFIG_FILE +
                         " or in the location given by the $CHANDLER_CONF_FILE environement variable\n")
        sys.exit(1)
else:
    config.read(os.environ['CHANDLER_CONF_FILE'])

# Get some variables from the configuration file
APIURI = config.get('oarapi', 'uri')
APILIMIT = config.get('oarapi', 'limit')
COLS = config.getint('output', 'columns')
NODENAME_REGEX = config.get('output', 'nodename_regex')
COL_SIZE = config.getint('output', 'col_size')
COL_SPAN = config.getint('output', 'col_span')
MAX_COLS = config.getint('output', 'max_cols')
USERS_STATS_BY_DEFAULT = config.getboolean('output', 'users_stats_by_default')
NODES_USAGE_BY_DEFAULT = config.getboolean('output', 'nodes_usage_by_default')
NODES_HEADER = config.get('output', 'nodes_header')
NODES_FORMAT = config.get('output', 'nodes_format')
try:
    COMMENT_PROPERTY = config.get('output', 'comment_property')
except:
    COMMENT_PROPERTY = ""
try:
    SEPARATIONS = config.get('output', 'separations')
except:
    SEPARATIONS = ""
SEPARATIONS = SEPARATIONS.split(',')
CACHING = config.getboolean('oarapi', 'caching')
if CACHING:
    CACHING_RESOURCES_FILE = config.get('oarapi', 'caching_resources_file')
    CACHING_RESOURCES_DELAY = config.getint(
        'oarapi', 'caching_resources_delay')
    CACHING_JOBS_FILE = config.get('oarapi', 'caching_jobs_file')
    CACHING_JOBS_DELAY = config.getint('oarapi', 'caching_jobs_delay')
else:
    CACHING_RESOURCES_FILE = ""
    CACHING_RESOURCES_DELAY = 0
    CACHING_JOBS_FILE = ""
    CACHING_JOBS_DELAY = 0

# Compute the number of columns depending on the COLUMNS environment variable
try:
    rows, columns = os.popen('stty size', 'r').read().split()
    COLS = int(int(columns) / COL_SIZE)
except:
    pass
if COLS == 0:
    COLS = 1
if COLS > MAX_COLS:
    COLS = MAX_COLS

# Options parsing
parser = OptionParser()
parser.add_option("-u", "--users",
                  action="store_true", dest="toggle_users", default=False,
                  help="Toggle printing users stats")
parser.add_option("-n", "--nodes",
                  action="store_true", dest="toggle_nodes", default=False,
                  help="Toggle printing nodes usage")
parser.add_option("-r", "--reload-cache",
                  action="store_true", dest="reload_cache", default=False,
                  help="Reload the cache")
(options, args) = parser.parse_args()

# Get rid of http_proxy if necessary
if config.getboolean('misc', 'ignore_proxy'):
    try:
        del os.environ['http_proxy']
    except:
        pass
    try:
        del os.environ['https_proxy']
    except:
        pass

# Functions


def get_from_api(uri):
    """
        Get an object from the api
    """
    headers = {'Accept': 'application/json'}
    r = requests.get(APIURI + uri + "?limit=" + APILIMIT, headers=headers)
    if r.status_code != 200:
        print("Could not get " + APIURI + uri)
        r.raise_for_status()
    if StrictVersion(requests.__version__) >= StrictVersion("2.0.0"):
        return r.json()
    else:
        return r.json


def get(uri, cache_file, cache_delay):
    """
        Get from the cache or from the api
    """
    if CACHING and not options.reload_cache and os.path.isfile(cache_file) and time.time() - os.path.getmtime(cache_file) < cache_delay:
        json_data = open(cache_file)
        return json.load(json_data)
    else:
        data = get_from_api(uri)
    if CACHING:
        chmod = True
        if os.path.isfile(cache_file):
            chmod = False
        file = open(cache_file, 'w')
        json.dump(data, file)
        file.close
        if chmod:
            os.chmod(cache_file, 0666)
    return data


def cprint(str, *args):
    """
        Custom print function to get rid of trailing newline and space
    """
    sys.stdout.write(str % args)

# Print a waiting message
print('Querying OAR API...\n\033[1A'),

# Get the data from the API
# TODO: paginated results management
resources = get(
    '/resources/details', CACHING_RESOURCES_FILE, CACHING_RESOURCES_DELAY)["items"]
jobs = get('/jobs/details', CACHING_JOBS_FILE, CACHING_JOBS_DELAY)["items"]

# Erase the waiting message
print("\033[2K")

# Compute assigned resources dictionnary
assigned_resources = {r["id"]: j["types"] for j in jobs
                      for r in j["resources"]
                      if r["status"] == "assigned"}

# Compute sorted node list
nodes = natsorted(set([r["network_address"] for r in resources]))

# Get the comment property if necessary
if COMMENT_PROPERTY != '':
    comment = {r["network_address"]: r[COMMENT_PROPERTY] for r in resources}

# Loop on nodes and resources
col = 0
down = 0
for node in nodes:
    c = 0
    node_resources = [r for r in resources if r["network_address"] == node]
    p = re.match(NODENAME_REGEX, node)
    node_str = p.group(1)
    if COMMENT_PROPERTY != '':
        node_str += " (" + comment[node] + ")"
    else:
        node_str += ": "
    string = node_str + " " * (COL_SPAN - len(node_str))
    cprint(Fore.RESET + Back.RESET + string)
    for r in node_resources:
        c += 1
        if r["state"] == "Dead":
            down += 1
            cprint(Back.RED + Fore.WHITE + "D")
        elif r["state"] == "Absent":
            if int(r["available_upto"]) > time.time():
                cprint(Back.CYAN + Fore.WHITE + " ")
            else:
                down += 1
                cprint(Back.RED + Fore.WHITE + "A")
        elif r["state"] == "Suspected":
            down += 1
            cprint(Back.RED + Fore.WHITE + "S")
        elif r["state"] == "Alive":
            try:
                types = assigned_resources[r["id"]]
            except:
                cprint(Back.GREEN + Fore.WHITE + " ")
            else:
                if "besteffort" in types:
                    cprint(Back.GREEN + Fore.BLACK + "B")
                elif "timesharing" in types:
                    cprint(Back.YELLOW + Fore.BLACK + "T")
                else:
                    cprint(Back.WHITE + Fore.BLACK + "J")
    cprint(Fore.RESET + Back.RESET)
    col += 1
    if col < COLS and node not in SEPARATIONS:
        cprint(" " * (COL_SIZE - COL_SPAN - c))
    else:
        col = 0
        print

# Legend
print(Fore.RESET + Back.RESET)
print
cprint(Back.GREEN + " " + Back.RESET + "=Free ")
cprint(Back.GREEN + Fore.BLACK + "B" +
       Back.RESET + Fore.RESET + "=Besteffort ")
cprint(Back.CYAN + " " + Back.RESET + "=Standby ")
cprint(Back.WHITE + Fore.BLACK + "J" + Back.RESET + Fore.RESET + "=Job ")
cprint(Back.RED + Fore.BLACK + "S" + Back.RESET + Fore.RESET + "=Suspected ")
cprint(Back.RED + Fore.BLACK + "A" + Back.RESET + Fore.RESET + "=Absent ")
cprint(Back.RED + Fore.BLACK + "D" + Back.RESET + Fore.RESET + "=Dead ")
print

# Reset terminal styles
print(Fore.RESET + Back.RESET + Style.RESET_ALL)

# Print summary
print "{} jobs, {} resources, {} down, {} used".format(len(jobs), len(resources), down, len(assigned_resources))


# Print users stats if necessary
if USERS_STATS_BY_DEFAULT ^ options.toggle_users and len(jobs) > 0:
    print
    user_resources = defaultdict(int)
    user_running = defaultdict(int)
    user_waiting = defaultdict(int)
    user_nodes = defaultdict(list)
    for j in jobs:
        if j["state"] == "Running" or j["state"] == "Finishing" or j["state"] == "Launching":
            user_running[j["owner"]] += 1
            user_resources[j["owner"]] += len(j["resources"])
            user_nodes[j["owner"]] += [n["network_address"]
                                       for n in j["nodes"]]
        elif j["state"] == "Waiting":
            user_waiting[j["owner"]] += 1
            user_resources[j["owner"]] += 0
            user_nodes[j["owner"]] += []

    print "               Jobs       Jobs"
    print "User          running    waiting   Resources    Nodes"
    print "====================================================="
    for u, r in user_resources.iteritems():
        nodes = set(user_nodes[u])
        print "{:<16} {:<10} {:<10} {:<10} {:<10}".format(u, user_running[u], user_waiting[u], r, len(nodes))

# Print nodes usage if necessary
if NODES_USAGE_BY_DEFAULT ^ options.toggle_nodes and len(jobs) > 0:
    print
    assigned_nodes = [[r["network_address"], j]
                      for j in jobs
                      for r in j["nodes"] if r["status"] == "assigned"]
    nodes_usage = defaultdict(list)
    for c in assigned_nodes:
        nodes_usage[c[0]].append(c[1])
    print NODES_HEADER
    for node in natsorted(nodes_usage.keys()):
        print node + ":"
        for job in nodes_usage[node]:
            d = time.strftime(
                "%H:%M:%S", time.gmtime(job["start_time"] + job["walltime"] - time.time()))
            r = [r for r in resources if r["network_address"] ==
                 node and r["id"] in [rj["id"] for rj in job["resources"]]]
            print "    " + NODES_FORMAT.format(job["id"], job["owner"], str(job["name"]), len(r), ",".join(job["types"]), d, job["project"])
